ledger/ledger-runner-common/src/main/scala/com/daml/ledger/runner/common/PureConfigReaderWriter.scala

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.runner.common
+
+import com.daml.ledger.api.tls.TlsVersion.TlsVersion
+import com.daml.ledger.api.tls.{SecretsUrl, TlsConfiguration, TlsVersion}
+import com.daml.lf.data.Ref
+import com.daml.lf.engine.EngineConfig
+import com.daml.lf.language.LanguageVersion
+import com.daml.lf.transaction.ContractKeyUniquenessMode
+import com.daml.lf.{VersionRange, interpretation, language}
+import com.daml.metrics.MetricsReporter
+import com.daml.platform.apiserver.SeedService.Seeding
+import com.daml.platform.apiserver.configuration.RateLimitingConfig
+import com.daml.platform.apiserver.{ApiServerConfig, AuthServiceConfig}
+import com.daml.platform.configuration.{
+  CommandConfiguration,
+  IndexServiceConfig,
+  InitialLedgerConfiguration,
+  PartyConfiguration,
+}
+import com.daml.platform.indexer.ha.HaConfig
+import com.daml.platform.indexer.{IndexerConfig, IndexerStartupMode}
+import com.daml.platform.services.time.TimeProviderType
+import com.daml.platform.store.DbSupport.{
+  ConnectionPoolConfig,
+  DataSourceProperties,
+  ParticipantDataSourceConfig,
+}
+import com.daml.platform.store.LfValueTranslationCache
+import com.daml.platform.store.backend.postgresql.PostgresDataSourceConfig
+import com.daml.platform.store.backend.postgresql.PostgresDataSourceConfig.SynchronousCommitValue
+import com.daml.platform.usermanagement.UserManagementConfig
+import com.daml.ports.Port
+import com.typesafe.config.ConfigValueFactory
+import io.netty.handler.ssl.ClientAuth
+import pureconfig.configurable.{genericMapReader, genericMapWriter}
+import pureconfig.error.CannotConvert
+import pureconfig.generic.semiauto._
+import pureconfig.{ConfigConvert, ConfigCursor, ConfigReader, ConfigWriter, ConvertHelpers}
+
+import scala.concurrent.duration.{Duration, FiniteDuration}
+import scala.jdk.DurationConverters.{JavaDurationOps, ScalaDurationOps}
+import scala.util.Try
+
+object PureConfigReaderWriter {
+  val Secret = "<REDACTED>"
+
+  /** Reads configuration object of `T` and `enabled` flag to find out if this object has values.
+    */
+  def optReaderEnabled[T](reader: ConfigReader[T]): ConfigReader[Option[T]] =
```

Comment:
I am not seeing the added benefit of the _enabled_ pattern but maybe I am missing something. Would like to discuss this

Reply:
It's not a benefit, it is a necessity, otherwise we should not be using an option in the config. With option - you cannot unset the value if it was setup by the higher configuration level.

Imagine you have a default set to a value:
```
api-server.rate-limit {
  queue-limit = 10
} 
```
which is modeled then in the case class with Option, i.e. if it is `Some` rate limit is enabled, otherwise - disabled:
```
case class ApiServerConfig(
...
  rateLimit: Option[RateLimitConfig]
...
```

You would not be able to disable rate-limit in your config file. With the feature implemented above - every such Optional, would be supporting `enabled` flag which can be then disabled like this:

```
api-server.rate-limit.enabled = false
```

Reply:
I was thinking about this. It seems more verbose and the handling logic more complex than other ideas that crossed my mind: I wonder whether a null override or a simple empty data struct (e.g. `{}`) wouldn't have worked as well while being simpler. Anyway, fine like this as well if you already discarded those 👍